packages/components/src/custom-select-control-v2/legacy-component/index.tsx

```diff
@@ -30,25 +30,22 @@ function CustomSelectControl( props: LegacyCustomSelectProps ) {
 	// Forward props + store from v2 implementation
 	const store = Ariakit.useSelectStore( {
 		async setValue( nextValue ) {
-			if ( ! onChange ) {
-				return;
-			}
+			if ( ! onChange ) return;
 
 			// Executes the logic in a microtask after the popup is closed.
 			// This is simply to ensure the isOpen state matches that in Downshift.
 			await Promise.resolve();
 			const state = store.getState();
 
+			const option = options.find( ( item ) => item.name === nextValue );
+
 			const changeObject = {
 				highlightedIndex: state.renderedItems.findIndex(
 					( item ) => item.value === nextValue
 				),
 				inputValue: '',
 				isOpen: state.open,
-				selectedItem: {
```

Comment:
The old object here didn't include any custom attributes that might was part of the original options array. Not sure we should instead pass these values through the Ariakit's API (e.g use `value` to store the style data), but selecting the option by name is a (hacky) workaround that works for now.

Reply:
cc @mirka

Reply:
OK, after spending more time with the codebase for V2/LegacyAdapter and the V1 component, I think I grasped what was wrong here. There were two issues:

1. The `selectedItem` object had its `name` and `key` both set to the `nextValue` argument passed to the `setValue` callback, resulting in both having the `name` attribute from the options array. This didn't look right.
2. The same structure was missing the style metadata (`style` and/or `className`) that was considered to be implicitly available in the item passed to the `onChange` callback in V1, which [some consumers expect](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/font-appearance-control/index.js).

I've fixed both in https://github.com/WordPress/gutenberg/pull/61272/commits/56461bdeac11bbd77a290d027bebc483076a89a6, but I'm wondering if this is the right approach and if should we keep the same API in V2 (as in keep the style metadata in the item passed to `onChange`?), or should we refactor the consumers to not rely on this self-contained data and instead move it to a store somewhere else? Of course, this would be work to be done in follow-up PRs.